### PR TITLE
support migration from intree driver to csi driver

### LINKED ISSUE #159: Test CSI migration functionality from in-tree driver to CSI driver
**Is your feature request related to a problem?/Why is this needed**

**Describe the solution you'd like in detail**

design doc: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #275: Cannot create PVC with volumesnapshot
**What happened**:
Step:
1. Install azuredisk-csi-driver successfully (#274)
2. create pvc successfully
3. create volumesnapshot/volumesnapshotcontents successfully.
4. create another pvc with volumesnapshot failed.

From log,  not sure why csi driver cannot get the resource, could you kindly help check if issue from csi driver side?  Thanks in advance.

Resource we have(snapshotcontents/snapshots):
$ kubectl get volumesnapshotcontents.snapshot.storage.k8s.io
NAME                                               AGE
snapcontent-9cfa4c37-8515-4375-8699-5435128827a2   54m
$ kubectl get volumesnapshots.snapshot.storage.k8s.io vs-03
NAME    AGE
vs-03   97m

Error message:
From snapshot container log:
_I0206 09:50:37.732505       1 reflector.go:161] Listing and watching *v1alpha1.VolumeSnapshotContent from github.com/kubernetes-csi/external-snapshotter/pkg/client/informers/externalversions/factory.go:117
E0206 09:50:37.734013       1 reflector.go:126] github.com/kubernetes-csi/external-snapshotter/pkg/client/informers/externalversions/factory.go:117: Failed to list *v1alpha1.VolumeSnapshotContent: the server could not find the requested resource (get volumesnapshotcontents.snapshot.storage.k8s.io)_
From controller container log:
_I0206 09:08:10.680201       1 controller.go:1199] provision "kube-system/restore-pvc-03" class "disk.csi.azure.com": started
I0206 09:08:10.691920       1 controller.go:398] skip translation of storage class for plugin: disk.csi.azure.com
I0206 09:08:10.692553       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"kube-system", Name:"restore-pvc-03", UID:"dbade117-3a31-4a4c-9c1c-a14d3b2972b0", APIVersion:"v1", ResourceVersion:"672665", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "kube-system/restore-pvc-03"
W0206 09:08:10.697522       1 controller.go:887] Retrying syncing claim "dbade117-3a31-4a4c-9c1c-a14d3b2972b0", failure 0
E0206 09:08:10.697555       1 controller.go:910] error syncing claim "dbade117-3a31-4a4c-9c1c-a14d3b2972b0": failed to provision volume with StorageClass "disk.csi.azure.com": error getting handle for DataSource Type VolumeSnapshot by Name vs-03: error getting snapshot vs-03 from api server: the server could not find the requested resource (get volumesnapshots.snapshot.storage.k8s.io vs-03)
I0206 09:08:10.697583       1 event.go:255] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"kube-system", Name:"restore-pvc-03", UID:"dbade117-3a31-4a4c-9c1c-a14d3b2972b0", APIVersion:"v1", ResourceVersion:"672665", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "disk.csi.azure.com": error getting handle for DataSource Type VolumeSnapshot by Name vs-03: error getting snapshot vs-03 from api server: the server could not find the requested resource (get volumesnapshots.snapshot.storage.k8s.io vs-03)_

**What you expected to happen**:
create  pvc with volumesnapshot successfully.

**How to reproduce it**:

**Anything else we need to know?**:

**Environment**:
- CSI Driver version: 0.6.0
- Kubernetes version (use `kubectl version`): 1.17.1
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools: MSI
- Others:

### PR DESCRIPTION
**Pull request by andyzhangx**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR 
 - upgrade csi-provisioner to 1.15.0 to support migration from intree driver to csi driver on k8s 1.17.0.
 - upgrade snapshot v1alpha1 to v1beta1, including
   - add [snapshot controller](https://github.com/kubernetes-csi/external-snapshotter/blob/master/README.md#usage)

**Which issue(s) this PR fixes**:

Fixes #159
Fixes #275

**Special notes for your reviewer**:
```
"--feature-gates=CSIMigration=true,CSIMigrationAzureDisk=true,CSIMigrationAzureFile=true"
```

```
$ kubectl describe pvc pvc-azuredisk1
...
Events:
  Type    Reason                 Age                From                                                                                               Message
  ----    ------                 ----               ----                                                                                               -------
  Normal  ExternalProvisioning   70s (x2 over 70s)  persistentvolume-controller                                                                        waiting for a volume to be created, either by external provisioner "disk.csi.azure.com" or manually created by system administrator
  Normal  Provisioning           70s                disk.csi.azure.com_csi-azuredisk-controller-6cf875f577-zxv87_6c096389-c76a-48db-979c-c55e84a9d002  External provisioner is provisioning volume for claim "default/pvc-azuredisk1"
  Normal  ProvisioningSucceeded  69s                disk.csi.azure.com_csi-azuredisk-controller-6cf875f577-zxv87_6c096389-c76a-48db-979c-c55e84a9d002  Successfully provisioned volume pvc-c4f05562-b1f1-4beb-8a93-0707bb9ec303
```

https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/

**Release note**:
```
support migration from intree driver to csi driver
```

### CODE REVIEW

**Comment by andyzhangx** [comment]
/hold
upgrade csi-provisioner would fail snapshot feature, need to investigate

**Comment by andyzhangx** [comment]
there is error when upgrade csi-provisioner from 1.4.0 to 1.5.0:
```
E0121 06:51:43.562754       1 controller.go:910] error syncing claim "aefea3db-d68d-4678-955d-089024c49b3b": failed to provision volume with StorageClass "disk.csi.azure.com": error getting handle for DataSource Type VolumeSnapshot by Name azuredisk-volume-snapshot: error getting snapshot azuredisk-volume-snapshot from api server: the server could not find the requested resource (get volumesnapshots.snapshot.storage.k8s.io azuredisk-volume-snapshot)
I0121 06:51:43.562819       1 event.go:281] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pvc-azuredisk-snapshot-restored", UID:"aefea3db-d68d-4678-955d-089024c49b3b", APIVersion:"v1", ResourceVersion:"1515900", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "disk.csi.azure.com": error getting handle for DataSource Type VolumeSnapshot by Name azuredisk-volume-snapshot: error getting snapshot azuredisk-volume-snapshot from api server: the server could not find the requested resource (get volumesnapshots.snapshot.storage.k8s.io azuredisk-volume-snapshot)
```

@ZeroMagic

**Comment by andyzhangx** [comment]
update: 
it's due to `csi-snapshotter` could not get lease, happens on 1.16+
```
I0121 07:50:27.379010       1 main.go:89] Version: v1.1.0-0-gac339597
I0121 07:50:27.460945       1 connection.go:151] Connecting to unix:///csi/csi.sock
I0121 07:50:32.411048       1 common.go:111] Probing CSI driver for readiness
I0121 07:50:32.413009       1 leaderelection.go:217] attempting to acquire leader lease  kube-system/external-snapshotter-leader-election...
```

**Comment by andyzhangx** [comment]
Update:
this PR works on AKS 1.17.0, looks like it only does not work aks-engine cluster, `external-snapshotter-leader-election` lease is not created on aks-engine cluster

**Comment by andyzhangx** [comment]
here is the solution, need to upgrade snapshot `v1alpha1` to `v1beta1`: https://github.com/kubernetes-csi/external-snapshotter/blob/master/README.md#upgrade-from-v1alpha1-to-v1beta1
It's not the lease issue.

**Comment by ZeroMagic** [comment]
We should follow the Usage. https://github.com/kubernetes-csi/external-snapshotter/blob/master/README.md#usage

`Common Snapshot Controller` should be installed in kube-system.

**Comment by ZeroMagic** [comment]
Then add some rules in `azuredisk-external-provisioner-role` and `azuredisk-external-snapshotter-role`.

In  `azuredisk-external-provisioner-role`
```
  - apiGroups: ["snapshot.storage.k8s.io"]
    resources: ["volumesnapshots"]
    verbs: ["get", "list"]
  - apiGroups: ["snapshot.storage.k8s.io"]
    resources: ["volumesnapshotcontents"]
    verbs: ["get", "list"]
```

In `azuredisk-external-snapshotter-role`
```
  - apiGroups: ["snapshot.storage.k8s.io"]
    resources: ["volumesnapshotcontents/status"]
    verbs: ["update"]
```

**Comment by ZeroMagic** [comment]
The file `deploy/example/snapshot/storageclass-azuredisk-snapshot.yaml` and `deploy/example/snapshot/azuredisk-volume-snapshot.yaml` should be modified.

```yaml
# deploy/example/snapshot/storageclass-azuredisk-snapshot.yaml
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshotClass
metadata:
  name: csi-azuredisk-vsc
driver: disk.csi.azure.com
deletionPolicy: Delete

```

```yaml
# deploy/example/snapshot/azuredisk-volume-snapshot.yaml
apiVersion: snapshot.storage.k8s.io/v1beta1
kind: VolumeSnapshot
metadata:
  name: azuredisk-volume-snapshot
spec:
  volumeSnapshotClassName: csi-azuredisk-vsc
  source:
    persistentVolumeClaimName: pvc-azuredisk
```

**Comment by andyzhangx** [comment]
also need to integrate new snapshot controller into csi driver, a lot work to do in this PR, it worked in my env now:
```
$ kubectl describe volumesnapshot azuredisk-volume-snapshot
Name:         azuredisk-volume-snapshot
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"snapshot.storage.k8s.io/v1beta1","kind":"VolumeSnapshot","metadata":{"annotations":{},"name":"azuredisk-volume-snapshot","n...
API Version:  snapshot.storage.k8s.io/v1beta1
Kind:         VolumeSnapshot
Metadata:
  Creation Timestamp:  2020-02-04T13:59:54Z
  Finalizers:
    snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
    snapshot.storage.kubernetes.io/volumesnapshot-bound-protection
  Generation:        1
  Resource Version:  48044
  Self Link:         /apis/snapshot.storage.k8s.io/v1beta1/namespaces/default/volumesnapshots/azuredisk-volume-snapshot
  UID:               2b0ef334-4112-4c86-8360-079c625d5562
Spec:
  Source:
    Persistent Volume Claim Name:  pvc-azuredisk
  Volume Snapshot Class Name:      csi-azuredisk-vsc
Status:
  Bound Volume Snapshot Content Name:  snapcontent-2b0ef334-4112-4c86-8360-079c625d5562
  Creation Time:                       2020-02-04T14:23:36Z
  Ready To Use:                        true
  Restore Size:                        10Gi
Events:                                <none>
```

File: test/e2e/driver/driver.go [hunk 1]
```diff
@@ -84,8 +84,8 @@ func getStorageClass(
 	}
 }
 
-func getVolumeSnapshotClass(generateName string, provisioner string) *v1alpha1.VolumeSnapshotClass {
```
Comment (andyzhangx) [inline test/e2e/driver/driver.go]: Hi @xing-yang I am trying to upgrade v1alpha1 to v1beta1, and always hit following error when running e2e test in creating snapshot class(test code is [here](https://github.com/andyzhangx/azuredisk-csi-driver/blob/0b119a046385dc2aa6474768637c2d965159fb76/test/e2e/testsuites/testsuites.go#L106)), while if I create snapshot by `kubectl` manually, it works well.
```
    Unexpected error:
        <*errors.StatusError | 0xc00043ec80>: {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "Failure",
                Message: "the server rejected our request due to an error in our request (post volumesnapshotclasses.snapshot.storage.k8s.io)",
                Reason: "Invalid",
                Details: {
                    Name: "",
                    Group: "snapshot.storage.k8s.io",
                    Kind: "volumesnapshotclasses",
                    UID: "",
                    Causes: [
                        {
                            Type: "UnexpectedServerResponse",
                            Message: "unknown",
                            Field: "",
                        },
                    ],
                    RetryAfterSeconds: 0,
                },
                Code: 422,
            },
        }
```

Comment (andyzhangx) [reply test/e2e/driver/driver.go]: do you know what's the possible reason? My k8s lib of csi driver is based on 1.15, is that the root cause, shall I upgrade to 1.17. 
Current situation is only snapshot e2e test failed on creating snapshot class, while manual snapshot e2e test works by using `kubectl`  
@xing-yang Thanks

Comment (xing-yang) [reply test/e2e/driver/driver.go]: What is the Kubernetes version you are running?  Do you have alpha version of CRDs installed on the test bed before?

Comment (andyzhangx) [reply test/e2e/driver/driver.go]: thanks, with help of @xing-yang this failure is due to `VolumeSnapshotClass.DeletionPolicy` is required in v1beta1 while it's optional in v1alpha1

**Comment by andyzhangx** [comment]
the namespace deletion timeout is due to storage class deleted before volume snapshot deletion, and then volume snapshot deletion will always fail:
```
I0207 05:01:54.225] Feb  7 05:01:54.225: INFO: deleting StorageClass azuredisk-5605-disk.csi.azure.com-dynamic-sc-bfp5t
I0207 05:01:54.259] Feb  7 05:01:54.258: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
I0207 05:01:54.323] STEP: Destroying namespace "azuredisk-5605" for this suite.
I0207 05:11:54.454] Feb  7 05:11:54.453: INFO: Waiting up to 30s for server preferred namespaced resources to be successfully discovered
I0207 05:11:55.042] Feb  7 05:11:55.042: INFO: namespace: azuredisk-5605, resource: volumesnapshots, items remaining: 1
I0207 05:11:55.631] Feb  7 05:11:55.631: INFO: namespace: azuredisk-5605, DeletionTimetamp: 2020-02-07 05:01:54 +0000 UTC, Finalizers: [kubernetes], Phase: Terminating
I0207 05:11:55.664] Feb  7 05:11:55.663: INFO: namespace: azuredisk-5605, total namespaces: 5, active: 4, terminating: 1
I0207 05:11:55.696] Feb  7 05:11:55.696: INFO: Couldn't delete ns: "azuredisk-5605": namespace azuredisk-5605 was not deleted with limit: timed out waiting for the condition, namespaced content other than pods remain (&errors.errorString{s:"namespace azuredisk-5605 was not deleted with limit: timed out waiting for the condition, namespaced content other than pods remain"})
```

**Pull request merged**